### PR TITLE
[Form] Add "reindex" option for the CollectionType

### PR DESCRIPTION
**Pull request by sergeyfedotov**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12936 |
| License | MIT |
| Doc PR | TODO |

Adds a new option to the collection field type. This option allows to reindex the data array if necessary.

TODO:
- [ ] submit changes to the documentation

### CODE REVIEW

**Comment by sergeyfedotov** [comment]
This PR needs to be reviewed again because the logic has been changed.

**Comment by stof** [comment]
This looks good to me, however I would prefer if @webmozart could review it as well.

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 1]
```diff
@@ -110,6 +116,29 @@ public function preSubmit(FormEvent $event)
             throw new UnexpectedTypeException($data, 'array or (\Traversable and \ArrayAccess)');
         }
 
+        // Fix array indices
+        if ($this->reindex) {
+            $originalData = $data;
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: What if $originalData is not an array? At the moment, this method also supports array-like objects, although I don't understand why. The check was introduced in 77bea81ed5d9513cd32e5ad2a530e945832ae3b9, but I don't know why anymore why. Any ideas @stof @Tobion?

**Comment by sergeyfedotov** [comment]
@webmozart I rewrote the code to support array-like objects.

**Comment by fabpot** [comment]
Why do we need a new option for that? Can't we always reindex?

**Comment by webmozart** [comment]
Thanks for your PR @sergeyfedotov! :) Unfortunately it's more complicated than that (suprise).

When we reindex the data submitted by the browser, we lose the ability to assign submitted data to entries of the collection. Imagine:

~~~php
$data = [
    0 => ['price' => 10],
    2 => ['price' => 20],
];

$collection = [
    0 => Product(id => 1, price => 10),
    1 => Product(id => 2, price => 12),
    2 => Product(id => 3, price => 15),
];
~~~

Instead of removing product with ID 2 from the list and updating product 3 to price 20, we now removed product 3 and updated product 2 (which may or may not be what we want to do). Similar issues are already described in https://github.com/symfony/symfony/issues/7468 and https://github.com/symfony/symfony/issues/7828.

A proper fix, from what I can judge right now without too much investigation, would involve the following steps:

* Add a new option `entry_name` (consistent with `choice_name`) that returns a string that uniquely identifies an entry of the collection and is used as name of the entry forms instead of the collection indices. This could be a property path (e.g. `id`) or a closure with the signature `function ($entry, $key)` where the default return value is `$key` to keep BC.

* On submit (`ResizeFormListener::preSubmit()`), rearrange the order of the fields to resemble the submitted data. This allows to reorder the collection on the client side.

* Create a custom data mapper that writes the collection based on equivalence of the field names and the calculated `entry_name` of each original collection entry. We cannot simply compare the indexes sent by the browser (the `entry_name` of each field) and in the collection (e.g. sequential integers - or not, depends on the implementation). The `if ($this->allowDelete) {` block in `ResizeFormListener::onSubmit()` should probably be moved to that data mapper.

* Add an option `reindex` (default `false`) that generates the property paths of the entry forms as sequential integer if set to `true`. Additionally, the data mapper must return the new collection with sequential indexes if set to `true`. This makes sure that error mapping works correctly, *as long as the domain model also keeps its collections sequentially indexed* (hence this is a BC break and optional).

* We should set `reindex` to `true` by default in Symfony 4. I expect this to become best practice in order to avoid the error mapping problems that we have right now.

**Comment by sergeyfedotov** [comment]
@webmozart thanks for the detailed answer. You are right that a proper solution is more complicated. This pull request only trying to solve a problem with the errors association. In fact is a data association problem.

I think that the PR may be closed?

**Pull request closed without merging**